ntplib.py

```diff
@@ -297,37 +300,62 @@ def request(self, host, version=2, port="ntp", timeout=5, address_family=socket.
         NTPStats object
         """
 
-        # lookup server address
-        addrinfo = socket.getaddrinfo(host, port, address_family)[0]
-        family, sockaddr = addrinfo[0], addrinfo[4]
-
-        # create the socket
-        sock = socket.socket(family, socket.SOCK_DGRAM)
-
+        dest_timestamp = None
         try:
-            sock.settimeout(timeout)
-
-            # create the request packet - mode 3 is client
-            query_packet = NTPPacket(
-                mode=3,
-                version=version,
-                tx_timestamp=system_to_ntp_time(time.time())
-            )
-
-            # send the request
-            sock.sendto(query_packet.to_data(), sockaddr)
-
-            # wait for the response - check the source address
-            src_addr = (None,)
-            while src_addr[0] != sockaddr[0]:
-                response_packet, src_addr = sock.recvfrom(256)
-
-            # build the destination timestamp
-            dest_timestamp = system_to_ntp_time(time.time())
-        except socket.timeout:
-            raise NTPException("No response received from %s." % host)
-        finally:
-            sock.close()
+            # lookup server address - request only datagram (i.e. UDP) addresses
+            addrinfo_a = socket.getaddrinfo(host, port, address_family, socket.SOCK_DGRAM)
+        except socket.gaierror:
+            # can happen for a v6 only resolvable hostname on a v4 only host (or a v4 only hostname on a v6 only host)
+            # getaddrinfo() tries to be clever; which is damn annoying!
+            raise NTPException("No resolved address(es) for %s." % host)
+
+        # we randomize the addresses in order to spread the load
+        # if a specific address does not work, no worries, we just continue to the next
+        # for sanity reasons (i.e. limiting random sample cpu usage); we limit the tries to _MAX_HOSTS_TO_TRY addresses
+        for addrinfo in random.sample(addrinfo_a, k=min(NTPClient._MAX_HOSTS_TO_TRY, len(addrinfo_a))):
+            if addrinfo[1] != socket.SOCK_DGRAM or addrinfo[2] != socket.IPPROTO_UDP:
+                # should not happen - no worries, try the next one
+                continue
+
+            family, sockaddr = addrinfo[0], addrinfo[4]
+
+            # create the socket for UDP datagram
+            try:
+                sock = socket.socket(family, socket.SOCK_DGRAM, socket.IPPROTO_UDP)
+            except OSError as e:
+                # should not happen - no worries, try the next one
+                continue
+
+            try:
+                # The timeout logic changes. We have N hosts to check, so we have to sum up the timeouts
+                sock.settimeout(timeout/min(NTPClient._MAX_HOSTS_TO_TRY, len(addrinfo_a)))
```

Comment:
I think a better way to do that would be to measure the timeout from the start of the function, i.e.:
```
deadline = time.time() + timeout
for addr in getaddrinfo():
    [...]
    sock.settimeout(max(0, deadline - time.time()))
```

This way the timeout is accurate no matter how many address we try, and whether they succeed or not.

Reply:
Correct. I like that. I can PR that. I think it needs an additional `if deadline - time.time() < 0` check followed by a `break` otherwise you could send a negative timeout to `socket.settimeout`[1].

[1] https://docs.python.org/3/library/socket.html?highlight=socket%20settimeout#socket.socket.settimeout "The value argument can be a nonnegative floating point number expressing seconds, or None."

Reply:
OK. This logic would work...
```
$ python
...
>>> import socket
>>> sock = socket.socket(socket.AF_INET, socket.SOCK_DGRAM, socket.IPPROTO_UDP)
>>> sock.settimeout(-1.0)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ValueError: Timeout value out of range
>>> 
```
That can handle the negative timeout.

Reply:
BTW: Yes, I see the `max(0,...)` however, if you pass 0 you would not get what you expect[1] as non-blocking mode would expect the response packet to already be there BEFORE the call to `sock.recvfrom()` which is unlikely. The code would loop needlessly sending NTP requests and ignoring the responses from past NTO requests. 

[1] If zero is given, the socket is put in non-blocking mode.

Reply:
Hear me out for a second. This new timeout code `deadline - time.time()` means that if the first returned address fails (with a timeout) then it's eaten up all the time and no other addresses are left to try because the timeout has been exhausted. There's no easy fix for this, as a simple backoff algorithm (lets say: dividing the remaining time by `2`) would simply bias the first returned address as having the lions-share of the timeout. I'll leave this as a to-do issue; as, once again, even in this case, it's emulating the existing `[0]` code.